### DISCUSSION TITLE
Changing methods

### DISCUSSION THREAD

**Discussion by lenticulariz** [BepInEx/BepInEx#898] [Help and Q&A]
Im making a mod using BepInEx and HarmonyX, i want to know how to change if statements in a method. I want to change the >= 4 in the first if statement, to >= 6. How would i do this? I've checked the documentation and dont know how i would.

`public void addSet(Transform obj)
	{
		ColorInputItem componentInParent = obj.GetComponentInParent<ColorInputItem>();
		if (componentInParent.numSets >= 4)
		{
			return;
		}
		if (componentInParent.myType == ColorInputItem.types.c)
		{
			foreach (ColorInputItem colorInput in colorInputs)
			{
				if (colorInput.myType == ColorInputItem.types.c)
				{
					colorInput.numSets++;
					colorInput.refreshSetBtns();
					localGColors[colorInput.ct].Add(colorInput.preview.color);
				}
			}
			changeSet(componentInParent, componentInParent.numSets);
			updateObjs("arena", componentInParent.currSet);
		}
		else
		{
			componentInParent.numSets++;
			componentInParent.refreshSetBtns();
			localBColors[componentInParent.bt].Add(componentInParent.preview.color);
			changeSet(componentInParent, componentInParent.numSets);
			updateObjs(obj.parent.name, componentInParent.currSet);
		}
		updateNumToggles();
	}`

**Comment by marcelbpunkt** [comment]
Almost 2 years late to the party but since no one ever ansered this, I'm doing it in case anyone else has a similar situation:
what you want to do is write a transpiler. Transpilers are methods that receive an `IEnumerable<CodeInstruction>` which is the method you want to patch as a sequence of IL code instructions (IL code is what C#/.NET code is compiled into) that you can manipulate before returning the modified sequence. If you know your way around assembler, Java byte code or anything similar, you won't have too much trouble navigating through IL code.
You can look at a method's IL code e.g. with ILSpy. At the top bar there's a dropdown menu saying "C#". It can be changed to "IL Code" or "IL code with C#" (recommended for easier orientation).
Now assuming  `componentInParent` is an instance of the type `UnityEngine.Component`, the code snippet in question should look similar to this:
```CIL
IL_0006: ldarg.0
IL_0008: ldfld class UnityEngine.Component OriginalClass::componentInParent
IL_0009: ldfld int32 Component::numSets
IL_000c: ldc.i4.4
IL_000f: blt IL_0042
IL_0010: ret

...

IL_0042: (code for when numSets < 4)

...
```
So in your transpiler you have to iterate over the `CodeInstruction` objects you get as an argument and replace the only `ldc.i4.4` instruction with `ldc.i4.6`. Assuming the method you want to patch is called `OriginalMethod` in the class `OriginalClass`, your transpiler would basically be this:

```C#
[HarmonyPatch(typeof(OriginalClass), nameof(OriginalClass.OriginalMethod)]
[HarmonyTranspiler]
static IEnumerable<CodeInstruction> MyTranspiler(IEnumerable<CodeInstruction> instructions)
{
    foreach (i in instructions)
    {
        if (i.OpCode == OpCodes.Ldc_I4_4) i.OpCode = OpCodes.Ldc_I4_6;
        yield return i;
    }
}
```
This ONLY works because the hard-coded 4 in the if statement is the only one in the whole method. If there were to be another hard-coded 4 in this method, that one would get changed to a 6 as well, so you would have to make further checks for the instructions before it.
You could also count the instructions and then just replace instruction at the correct position but this approach is much more likely to break with the next game update.

See also:
[Harmony Wiki - Transpilers](https://harmony.pardeike.net/articles/patching-transpiler.html)
[List of CIL instructions](https://en.wikipedia.org/wiki/List_of_CIL_instructions)
